docs/adr/0005-remote-settings-client.md

```diff
@@ -0,0 +1,183 @@
+* There’s no requirement for any kind of scheduling or awareness of network state - when we are requested for content, we do it immediately and return an appropriate error if it can not be fetched.
+* There’s no requirement to support publishing records, requesting reviews or providing approvals via this new library.
+* There’s no requirement that push be used to communicate changes to the application (eg, to enable rapid-enrolment type features)
+* There’s no requirement to manage buckets, groups and collections via this new library.
+
+### Initial Requirements
+The requirements we do have for the initial version are:
+* The library should allow fetching records from Mozilla’s Remote Settings servers. This includes support for attachments, and fetching [incremental changes](https://docs.kinto-storage.org/en/latest/api/1.x/filtering.html#polling-for-changes).
+* The library should not create threads or run any event loops - the mobile apps themselves are responsible for all threading requirements. While this might change in the future, considering this kind of change to our mobile applications is out of scope for this project.
+* We must use Necko for all networking on Android, must enforce all connections are via valid https hosts (although some test-only exceptions might be helpful for QA, such as allowing localhost connections to be http)
+* The library should be the only remote-settings library used in the browser. Specifically, this means that Nimbus must also be capable of using the library, and the work to move Nimbus to the library must be considered as part of the project.
+
+## Existing Libraries
+We have identified the following libraries which may be suitable for this project.
+### Remote-settings on desktop
+There is a version of the [remote settings client in desktop](https://searchfox.org/mozilla-central/source/services/settings/RemoteSettingsClient.jsm), written in Javascript.
+It has been used and been relatively stable since at least 2018, so can be considered very capable, but the roadblock to it being suitable for use by our mobile browsers is that it is written in Javascript, so while it might be possible to expose it to Android via geckoview, there’s no reasonable path to have it made available to iOS.
+
+### Rust Remote Settings Client
+There is an existing [remote settings client on github](https://github.com/mozilla-services/remote-settings-client).
+This client is written in Rust and has evolved over a number of years.
+The most recent changes were made to support being used in [Merino](https://github.com/mozilla-services/merino), which was re-written in Python, so there are no known consumers of this library left.
+
+The main attributes of this library relevant to this discussion are:
+* It’s written in Rust, but has no FFI - ie, it’s currently only consumable by other Rust code.
```

Comment:
Thanks for commenting and I'm sorry to hear you are disappointed with the outcome. Something very relevant that the document doesn't mention is our timelines and resourcing. This work will be carried out by the mobile team, who need this in Q1, and have limited experience with Rust. To be frank, we really were looking for the simplest strategy that would meet our requirements, and offer us a path forward should those requirements change.

The point about signature verification being added for Nimbus is illuminating - Nimbus is well over 2 years old. I can say with absolute certainty that if the existing client was already being used by Nimbus, that is what we would have used - the decision is really "let's use what Nimbus uses" rather than "let's not use the remote settings client". As far as I'm aware, there's no plan to move Nimbus to this client, and we don't think it's reasonable to make that a pre-requisite for this effort when it's clear Nimbus doesn't actually require it to happen and we really don't have the resources to make it happen. The fact no resources have been found in the last 2 years to make that happen was a key factor in sticking with the status quo.

If we did end up requiring signature validation etc, I'm quite confident that we would learn from and directly use as much as possible from that client. Our decision isn't a reflection on the quality of the client, but rather a pragmatic decision based on actual requirements and resourcing. Further, it's not clear that this work would make it significantly more difficult to move to a different client implementation in the future than it is now - the fact we have 2 consumers of the same client instead of one shouldn't make it twice as difficult.

I hope that helps understand where we are coming from, and hope that you will participate with reviews and planning when requirements get more sophisticated - as mentioned above, I don't see any reason this effort would make it much more effort to take the new client in the future than it would now. Please let us know if we can answer any other questions.